`ares_crypto`

matthew-levan opened the pull request:
This PR removes our dependency on `urcrypt` for binary builds by replacing the `urcrypt` functions we used for our currently-implemented crypto jets (`aes_siv`, `ed25519`, and `sha` jets) with our own Rust-native functions. Resolves #156.

TODO:
- [x] make sure none of the functions allocate (`alloc` features are disabled in all dependency crates)
- [x] `ed`
  - [x] `puck`
  - [x] `shar`
  - [x] `sign`
  - [x] `veri`
- [x] `aes`
  - [x] `siv{a,b,c}_en`
  - [x] `siv{a,b,c}_de`
- [x] replace `urcrypt` functions with `ares_crypto` in all jets
- [x] build and boot to `dojo` without `urcrypt`
- [x] ensure zero-length messages are handled correctly

eamsden commented on rust/ares_crypto/src/ed25519.rs:
```diff
@@ -1,38 +1,84 @@
 #[cfg(test)]
 #[cfg(feature = "test_vs_urcrypt")]
-mod ucrypt_test {
-    // tests comparing urcrypt and ares_crypto output for ed25519 functions go here
```
please leave this comment in and extend it with the cargo command to run these tests, since a regular `cargo test` will not run them

eamsden commented on rust/ares/src/noun.rs:
```diff
@@ -249,6 +249,11 @@ impl DirectAtom {
         let bytes: &[u8; 8] = unsafe { std::mem::transmute(&self.0) };
         &bytes[..]
     }
+
+    pub fn as_mut_bytes(&mut self) -> &mut [u8] {
+        let bytes: &mut [u8; 8] = unsafe { std::mem::transmute(&mut self.0) };
+        &mut bytes[..]
+    }
```
No.

- `transmute` is banned
- This will return a mutable reference to the existing direct atom, causing the atom to be modified by any writes through the mutable reference.

If this is being used to construct a direct atom by creating a local initialized to `D(0)` and mutating it, then there are a couple of alternatives:

1. Create a mutable local of type `[u8; 8]` and, when it is properly initialized, call `u64::from_le_bytes()` and initialize the atom from there
2. Create an indirect atom using `IndirectAtom::new_raw_mut_bytes()` and call `normalize_to_atom()` afterwards.

The second approach is always preferred because it ensures we do not accidentally construct a floating indirect atom or cell by overflowing the 63 bits of a direct atom.

eamsden reviewed (changes requested):
Actually looks pretty good, but I tagged a couple of issues.

eamsden commented on rust/ares/src/noun.rs:
```diff
@@ -806,6 +824,14 @@ impl Atom {
         }
     }
 
+    pub fn as_mut_bytes(&mut self) -> &mut [u8] {
+        if self.is_direct() {
+            unsafe { self.direct.as_mut_bytes() }
+        } else {
+            unsafe { self.indirect.as_mut_bytes() }
+        }
+    }
+
```
No. See above.

eamsden reviewed (changes requested):
Using `as_mut_bytes()` on atoms will cause these atoms to be mutated in place, breaking referential transparency. Please correct all instances of this.

matthew-levan commented:
> Using `as_mut_bytes()` on atoms will cause these atoms to be mutated in place, breaking referential transparency. Please correct all instances of this.

Done. Now working on ensuring zero-length messages are handled correctly.

matthew-levan commented:
@eamsden I ripped out the `as_mut_bytes` implementation from `DirectAtom` and replaced instances where it was used with new `IndirectAtom` allocations which I copy data into. I've also updated jets to handle zero-lengths correctly and added tests to demonstrate. This is ready for another review. The code isn't the cleanest, but I'm tagging you here for expediency nonetheless.

eamsden commented on rust/ares/src/jets/lock/aes.rs:
```diff
@@ -185,59 +159,42 @@ mod util {
-            let siv_data = _allocate_atoms(stack, ads)?;
+            let ac_siv_data = _allocate_ads(stack, ads)?;
+            let siv_data: &mut [&mut [u8]] = std::slice::from_raw_parts_mut(
+                ac_siv_data.as_mut_ptr() as *mut &mut [u8],
```
why is this doubly indirected?

matthew-levan replied:
Good question, I'm not sure exactly what's going on here or why it works. How are the lengths of the byte buffers in `AcAesSivData` structs being translated to the lengths of the byte slices in `siv_data` here? By constructing the `siv_data` variable as we do here, are we getting a slice of byte slices whose elements are `[bytes, length]`? If that's the caes, I'm not sure that's desirable. The tests (taken from both IETF test cases and `dojo` examples) do pass, though.

matthew-levan replied:
@ashelkovnykov would you mind taking a look at this as well?

ashelkovnykov replied:
This is correct:

The input is a list of atoms which are converted into a contiguous array of `AcAesSivData` objects, which is then used by the encoder. We pass it from the jets code as an array of byte arrays, because we need to reverse the byte ordering prior to encoding for each element. When used by the encoder, it implicitly converts our input to just a byte buffer. Vere does something similar.

We could make the byte ordering reversal explicit in the jet code and then pass just a byte buffer, if we think that's best. However, that's a different discussion than correctness.

eamsden commented on rust/ares/src/noun.rs:
```diff
@@ -474,6 +483,10 @@ impl IndirectAtom {
         unsafe { from_raw_parts(self.data_pointer() as *const u8, self.size() << 3) }
     }
 
+    pub fn as_mut_bytes(&mut self) -> &mut [u8] {
+        unsafe { from_raw_parts_mut(self.data_pointer_mut() as *mut u8, self.size() << 3) }
+    }
+
```
this should not exist. If a noun has already been created then we should never mutate it. If we need to write into a mutable buffer to make a noun then allocate a new one with `IndirectAtom::new_raw_mut_bytes`

eamsden reviewed (changes requested):
No `as_mut_ptr()` for atoms!

matthew-levan commented:
`_allocate_ads` means "allocate associated datas", not "allocate atoms". I
clarify this to note that we aren't using `ac_siv_data.as_mut_ptr()` to get
a mutable pointer to an atom. Does that make sense?

On Wed, Jan 17, 2024 at 5:36 PM Edward Amsden ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> No as_mut_ptr() for atoms!
> ------------------------------
>
> In rust/ares/src/jets/lock/aes.rs
> <https://github.com/urbit/ares/pull/189#discussion_r1456571133>:
>
> >      ) -> Result {
>          unsafe {
> -            let siv_data = _allocate_atoms(stack, ads)?;
> +            let ac_siv_data = _allocate_ads(stack, ads)?;
> +            let siv_data: &mut [&mut [u8]] = std::slice::from_raw_parts_mut(
> +                ac_siv_data.as_mut_ptr() as *mut &mut [u8],
>
> why is this doubly indirected?
> ------------------------------
>
> In rust/ares/src/noun.rs
> <https://github.com/urbit/ares/pull/189#discussion_r1456573596>:
>
> > +    pub fn as_mut_bytes(&mut self) -> &mut [u8] {
> +        unsafe { from_raw_parts_mut(self.data_pointer_mut() as *mut u8, self.size() << 3) }
> +    }
> +
>
> this should not exist. If a noun has already been created then we should
> never mutate it. If we need to write into a mutable buffer to make a noun
> then allocate a new one with IndirectAtom::new_raw_mut_bytes
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/urbit/ares/pull/189#pullrequestreview-1828329346>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AV2DQRH4OWUYQRJ5FKW2DDDYPBG5ZAVCNFSM6AAAAABAYWSHKKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQMRYGMZDSMZUGY>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

eamsden commented:
> `_allocate_ads` means "allocate associated datas", not "allocate atoms". I clarify this to note that we aren't using `ac_siv_data.as_mut_ptr()` to get a mutable pointer to an atom. Does that make sense?


that makes sense, but the PR still adds an `as_mut_ptr()` method for indirect atoms which simply should not exist.

matthew-levan commented:
Hm maybe it did at one time but I've since removed it. The only
modification to `noun.rs` now is the addition of `IndirectAtom::
new_raw_mut_bytearray`.

On Wed, Jan 17, 2024 at 11:02 PM Edward Amsden ***@***.***>
wrote:

> _allocate_ads means "allocate associated datas", not "allocate atoms". I
> clarify this to note that we aren't using ac_siv_data.as_mut_ptr() to get
> a mutable pointer to an atom. Does that make sense?
>
> that makes sense, but the PR still adds an as_mut_ptr() method for
> indirect atoms which simply should not exist.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/urbit/ares/pull/189#issuecomment-1897754578>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AV2DQRHN3PVQTIPKTW6NYELYPCNFTAVCNFSM6AAAAABAYWSHKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOJXG42TINJXHA>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

ashelkovnykov commented on rust/ares/src/jets/lock/sha.rs:
```diff
@@ -119,6 +143,14 @@ mod tests {
     fn test_shas() {
         let c = &mut init_context();
 
+        let sam = T(&mut c.stack, &[D(1), D(0)]);
+        assert_jet_ubig(
+            c,
+            jet_shas,
+            sam,
+            ubig!(_0x4abac214e1e95fe0c60df79d09cbd05454a4cb958683e02318aa147f2a5e6d60),
+        );
+
```
Do these tests have any significance? Or are they based on dojo calls from a ship running on Vere?

Theoretically, the NIST is supposed to have standardized test cases, but both of the links I've found are returning a 503...
[1](https://csrc.nist.gov/groups/ST/toolkit/examples.html)
[2](https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Standards-and-Guidelines/documents/examples/SHA256.pdf)

ashelkovnykov commented on rust/ares_crypto/src/ed25519.rs:
```diff
@@ -0,0 +1,212 @@
+
+    #[test]
+    fn test_ed_sign() {
+        // from https://datatracker.ietf.org/doc/html/rfc8032#section-7.1
+        // let msg: [u8; 0] = [];
+        // let seed_src = &ubig!(_0x9d61b19deffd5a60ba844af492ec2cc44449c5697b326919703bac031cae7f60)
+        //     .to_le_bytes();
+        // let mut seed: [u8; 32] = [0; 32];
+        // seed.copy_from_slice(seed_src);
+        // let mut signature: [u8; 64] = [0; 64];
+        // ac_ed_sign(&msg, &seed, &mut signature);
+        // let signature_src = &ubig!(_0xe5564300c360ac729086e2cc806e828a84877f1eb8e5d974d873e065224901555fb8821590a33bacc61e39701cf9b46bd25bf5f0595bbe24655141438e7a100b).to_le_bytes();
+        // let mut test_signature: [u8; 64] = [0; 64];
+        // test_signature.copy_from_slice(signature_src);
+        // assert_eq!(signature, test_signature);
```
Cruft left over from getting `ares_crypto` to work?

ashelkovnykov commented on rust/ares_crypto/src/ed25519.rs:
```diff
@@ -0,0 +1,212 @@
+        );
+    }
+
+    // #[test]
+    // fn test_ed_shar() {
+    // }
```
Cruft left over from getting `ares_crypto` to work?

ashelkovnykov replied:
Ah, nevermind - I see now. It's just a stub for a test that's yet unimplemented (unlike the other functions in this file, `shar` is currently only tested in relation to `urcrypt`).

ashelkovnykov commented on rust/ares/src/jets/lock/ed.rs:
```diff
@@ -142,6 +132,16 @@ mod tests {
     fn test_puck() {
         let c = &mut init_context();
 
+        let sam = A(
+            &mut c.stack,
+            &ubig!(_0x0),
+        );
+        let ret = A(
+            &mut c.stack,
+            &ubig!(_0x29da598ba148c03aa643e21d77153265730d6f2ad0a8a3622da4b6cebc276a3b),
+        );
+        assert_jet(c, jet_puck, sam, ret);
+
```
Is there any significance to these tests, or are they just from the dojo of a ship running on Vere?

matthew-levan replied:
The significance is that they are tests from a real Vere `dojo`, yes. Future work to add test vectors from RFC 8032 is noted in the above comment.

ashelkovnykov commented on rust/ares/src/jets/lock/ed.rs:
```diff
@@ -47,17 +44,17 @@ pub fn jet_shar(context: &mut Context, subject: Noun) -> Result {
 
         public[0..pub_bytes.len()].copy_from_slice(pub_bytes);
-        secret[0..sec_bytes.len()].copy_from_slice(sec_bytes);
+        secret[0..pub_bytes.len()].copy_from_slice(sec_bytes);
```
copy-paste error

matthew-levan replied:
Not sure what you mean here. Will you clarify please?

matthew-levan replied:
Ah, nevermind!

matthew-levan commented on rust/ares/src/jets/lock/aes.rs:
```diff
@@ -377,6 +334,22 @@ mod tests {
     pub fn test_sivb_en() {
         let c = &mut init_context();
 
+        /*
+        > (~(en sivb:aes:crypto [key=0x0 vec=~]) txt=0x0)
+        [p=0x8fb.4085.a9b9.3662.ab44.f911.e47e.9ccd q=0 r=0x0]
+         */
+        fn sample(_s: &mut NockStack) -> Noun {
+            D(0)
+        }
+        fn context(s: &mut NockStack) -> Noun {
+            let sample = T(s, &[D(0), D(0)]);
+            T(s, &[D(0), sample, D(0)])
+        }
+
+        let siv = A(&mut c.stack, &ubig!(0x8fb4085a9b93662ab44f911e47e9ccd));
+        let res = T(&mut c.stack, &[siv, D(0), D(0x0)]);
+        assert_jet_in_door(c, jet_sivb_en, &[sample], &[context], res);
+
```
It's just one I got from the `dojo`.

matthew-levan commented on rust/ares/src/jets/lock/aes.rs:
```diff
@@ -404,6 +377,22 @@ mod tests {
     pub fn test_sivc_en() {
         let c = &mut init_context();
 
+        /*
+        > (~(en sivc:aes:crypto [key=0x0 vec=~]) txt=0x0)
+        [p=0x2c6a.abc5.bb25.1140.e221.d70b.fb31.c519 q=0 r=0x0]
+         */
+        fn sample(_s: &mut NockStack) -> Noun {
+            D(0)
+        }
+        fn context(s: &mut NockStack) -> Noun {
+            let sample = T(s, &[D(0), D(0)]);
+            T(s, &[D(0), sample, D(0)])
+        }
+
+        let siv = A(&mut c.stack, &ubig!(0x2c6aabc5bb251140e221d70bfb31c519));
+        let res = T(&mut c.stack, &[siv, D(0), D(0x0)]);
+        assert_jet_in_door(c, jet_sivc_en, &[sample], &[context], res);
+
```
Same answer as above.

matthew-levan commented on rust/ares_crypto/src/sha.rs:
```diff
@@ -0,0 +1,118 @@
+    message.reverse();
+    let mut hasher = Sha1::new();
+    hasher.update(message);
+    let mut result = hasher.finalize();
```
Nice, good catch.

matthew-levan commented on rust/ares/src/noun.rs:
```diff
@@ -436,6 +436,15 @@ impl IndirectAtom {
+    pub unsafe fn new_raw_mut_bytearray<'a, const N: usize, A: NounAllocator>(
+        allocator: &mut A,
+    ) -> (Self, &'a mut [u8; N]) {
+        let word_size = (std::mem::size_of::<[u8; N]>() + 7) << 3;
```
I think you're right, though-- take, for example, `N` is `8`. In the case of a left shift (as written), the result is `64`, not `1` as it should be.

ashelkovnykov commented:
This PR is almost ready to ship.

Minimum necessary changes:
- [byte array allocation size bug](https://github.com/urbit/ares/pull/189#discussion_r1472702826)
- [SHA message slicing bug](https://github.com/urbit/ares/pull/189#discussion_r1474046046)

Changes I'd prefer to have, but aren't necessary to merge:
- [singleton `hasher`s](https://github.com/urbit/ares/pull/189#discussion_r1472719317)
- Implement test cases from well known sources (e.g. RFCs, NIST, etc.). We can get away without these, methinks, since we're now relying on Rust crates which are very well tested - but better safe than sorry!

matthew-levan reviewed (commented):
Thanks for the review @ashelkovnykov. I pushed a couple commits to address the byte array allocation size bug and the SHA message slicing bug. With respect to the single `hasher` use in our SHA module, where would the singleton be initialized? A global? Passed in to the functions via the caller? I propose merging ASAP and circling back to use the singleton pattern afterwards.

matthew-levan commented on rust/ares/src/jets/lock/sha.rs:
```diff
@@ -17,19 +17,22 @@ pub fn jet_shas(context: &mut Context, subject: Noun) -> Result {
+
+        let msg_len = met(3, ruz);
+        if msg_len > 0 {
+            let (_msg_ida, message) = IndirectAtom::new_raw_mut_bytes(stack, msg_len);
+            message.copy_from_slice(&ruz.as_bytes()[0..msg_len]);
```
We definitely can @ashelkovnykov, a pretty easy job too.

matthew-levan commented:
I removed `nix` from the CI in the latest commit, so it will need to be reverted and the CI error will need to be resolved. I wasn't able to fix the CI error quickly without removing `nix`.

ashelkovnykov commented:
@matthew-levan That's okay, @eamsden and I are going to look at fixing the Nix build using `musl`
